pallets/asset-registry/src/lib.rs

```diff
@@ -265,181 +323,170 @@ pub mod pallet {
+				let bounded_name = Self::to_bounded_name(name)?;
+				Some(bounded_name)
+			} else {
+				None
+			};
 
-			if let Some(loc) = location {
-				ensure!(asset_id != T::NativeAssetId::get(), Error::<T>::CannotUpdateLocation);
-				ensure!(
-					Self::location_assets(&loc).is_none(),
-					Error::<T>::LocationAlreadyRegistered
-				);
-				AssetLocations::<T>::insert(asset_id, &loc);
-				LocationAssets::<T>::insert(&loc, asset_id);
+			let bounded_symbol = if let Some(symbol) = symbol {
+				Some(Self::to_bounded_name(symbol)?)
+			} else {
+				None
+			};
 
-				Self::deposit_event(Event::LocationSet {
-					asset_id,
-					location: loc,
-				});
-			}
+			let details = AssetDetails::new(
+				bounded_name,
+				asset_type,
+				existential_deposit.unwrap_or(DEFAULT_ED),
+				bounded_symbol,
+				decimals,
+				xcm_rate_limit,
+				is_sufficient,
+			);
 
+			Self::do_register_asset(asset_id, &details, location)?;
 			Ok(())
 		}
 
 		/// Update registered asset.
 		///
-		/// Updates also mapping between name and asset id if provided name is different than currently registered.
+		/// All parameteres are optional and value is not updated if param is `None`.
+		///
+		/// `decimals` - can be update by `UpdateOrigin` only if it wasn't set yet. Only
+		/// `RegistryOrigin` can update `decimals` if it was previously set.
+		///
+		/// `location` - can be updated only by `RegistryOrigin`.
 		///
 		/// Emits `Updated` event when successful.
-
-		// TODO: No tests
 		#[pallet::call_index(1)]
 		#[pallet::weight(<T as Config>::WeightInfo::update())]
 		pub fn update(
 			origin: OriginFor<T>,
 			asset_id: T::AssetId,
-			name: Vec<u8>,
-			asset_type: AssetType<T::AssetId>,
-			existential_deposit: Option<T::Balance>,
-			xcm_rate_limit: Option<T::Balance>,
+			name: Option<Vec<u8>>,
+			asset_type: Option<AssetType<T::AssetId>>,
+			existential_deposit: Option<Balance>,
+			xcm_rate_limit: Option<Balance>,
+			is_sufficient: Option<bool>,
+			symbol: Option<Vec<u8>>,
+			decimals: Option<u8>,
+			location: Option<T::AssetNativeLocation>,
 		) -> DispatchResult {
-			T::RegistryOrigin::ensure_origin(origin)?;
+			let is_registry_origin = match T::RegistryOrigin::ensure_origin(origin.clone()) {
+				Ok(_) => true,
+				Err(_) => {
+					T::UpdateOrigin::ensure_origin(origin)?;
+
+					false
+				}
+			};
 
 			Assets::<T>::try_mutate(asset_id, |maybe_detail| -> DispatchResult {
-				let mut detail = maybe_detail.as_mut().ok_or(Error::<T>::AssetNotFound)?;
+				let mut details = maybe_detail.as_mut().ok_or(Error::<T>::AssetNotFound)?;
 
-				let bounded_name = Self::to_bounded_name(name)?;
+				let new_bounded_name = if let Some(n) = name {
+					let new_name = Self::to_bounded_name(n)?;
+					ensure!(Self::asset_ids(&new_name).is_none(), Error::<T>::AssetAlreadyRegistered);
+
+					if let Some(old_name) = &details.name {
+						AssetIds::<T>::remove(old_name);
+					}
+
+					if Some(new_name.clone()) != details.name {
```

Comment:
As I read this, it will remove the name if you send none in the function but the docs for it says it doesn't get updated.

Reply:
I don't think so, this whole block is executed only if name is not `None`. 
There is this test:
https://github.com/galacticcouncil/HydraDX-node/blob/1156f597f469731e9b9e64a6dc0dd2978499b4a0/pallets/asset-registry/src/tests/update.rs#L170C1-L224C2

I can write another one if you have specific scenario in mind.

Reply:
> I think the name is removed if it is the same as the previous one

You can't name to previous one. It will fail on this check
https://github.com/galacticcouncil/HydraDX-node/blob/cdfc19aca97d1e0c011d28f50646ee1c07e78c3b/pallets/asset-registry/src/lib.rs#L402